vortex-ipc/src/io/object_store.rs

```diff
@@ -0,0 +1,103 @@
+use std::future::Future;
+use std::io::Cursor;
+use std::ops::Range;
+
+use bytes::{Bytes, BytesMut};
+use object_store::path::Path;
+use object_store::{ObjectStore, PutPayload};
+use vortex_buffer::io_buf::IoBuf;
+use vortex_buffer::Buffer;
+use vortex_error::VortexResult;
+
+use crate::io::{VortexRead, VortexReadAt, VortexWrite};
+
+pub trait ObjectStoreExt {
+    fn vortex_read(
+        &self,
+        location: &Path,
+        range: Range<usize>,
+    ) -> impl Future<Output = VortexResult<impl VortexRead>>;
+
+    fn vortex_reader(&self, location: &Path) -> impl VortexReadAt;
+
+    fn vortex_writer(&self, location: &Path) -> impl VortexWrite;
+}
+
+impl<O: ObjectStore> ObjectStoreExt for O {
+    async fn vortex_read(
+        &self,
+        location: &Path,
+        range: Range<usize>,
+    ) -> VortexResult<impl VortexRead> {
+        let bytes = self.get_range(location, range).await?;
+        Ok(Cursor::new(Buffer::Bytes(bytes)))
+    }
+
+    fn vortex_reader(&self, location: &Path) -> impl VortexReadAt {
+        ObjectStoreReadAt::new(self, location)
+    }
+
+    fn vortex_writer(&self, location: &Path) -> impl VortexWrite {
+        ObjectStoreWriter::new(self, location)
+    }
+}
+
+pub struct ObjectStoreReadAt<'a, 'b, O: ObjectStore> {
+    object_store: &'a O,
+    location: &'b Path,
+}
+
+impl<'a, 'b, O: ObjectStore> ObjectStoreReadAt<'a, 'b, O> {
+    pub fn new(object_store: &'a O, location: &'b Path) -> Self {
+        Self {
+            object_store,
+            location,
+        }
+    }
+}
+
+impl<'a, 'b, O: ObjectStore> VortexReadAt for ObjectStoreReadAt<'a, 'b, O> {
+    async fn read_at_into(&self, pos: u64, mut buffer: BytesMut) -> std::io::Result<BytesMut> {
+        let start_range = pos as usize;
+        let bytes = self
+            .object_store
+            .get_range(self.location, start_range..(start_range + buffer.len()))
+            .await?;
+        buffer.as_mut().copy_from_slice(bytes.as_ref());
```

Comment:
this copy is unfortunate

Reply:
hmm yea, kinda annoying. opendal seems to be the same. i suppose given that tls is happening (in most cases) you can't truly do zero-copy from the NIC anyway, but still

Reply:
You can avoid copying again though. I haven't dug too deep but they must be getting buffer form somewhere, probably allocated by the http client, and ideally we'd be able to pass it down. But, all of that is a lot of custom logic for now

Reply:
yea it's using reqwest deep under the hood, to my knowledge I don't think they let you pass a set of buffers to read payload data into